nbt-utils/nbt_utils.cpp

```diff
@@ -270,6 +276,422 @@ wr_payload(LongArrayTag) {
+                  c += 'a' - 'A';
+                switch (c) {
+                  case 'b':
+                    if (!hadDot)
+                      types |= tagTypeToMask(TagType::Byte);
+                    break;
+                  case 's':
+                    if (!hadDot)
+                      types |= tagTypeToMask(TagType::Short);
+                    break;
+                  case 'l':
+                    if (!hadDot)
+                      types |= tagTypeToMask(TagType::Long);
+                    break;
+                  case 'f':
+                    types |= tagTypeToMask(TagType::Float);
+                    break;
+                  default:
+                    stream.setstate(std::ios_base::goodbit);  // FIXME EOF handling doesn't seem to behave correctly
+                    types |= tagTypeToMask(TagType::Double);
+                    if (!hadDot)
+                      types |= tagTypeToMask(TagType::Int);
+                    break;
+                }
+                break;
+              }
+              c = stream.get();
+            } while (true);
+          }
+          goto end;
+        }
+    }
+  }
+end:
+  stream.seekg(savedPos);
+  return types;
+}
+
+#define wr_snbt(klass) template<> void nbt::klass::writeSNBT(std::ostream &stream) const
+#define rd_snbt(klass) template<> bool nbt::klass::readSNBT(std::istream &stream) 
+ 
+#define rd_snbt_suffix(tagCls, readType, suffixUpper, suffixLower) \
+  rd_snbt(tagCls) { \
+    std::string buf; \
+    char c; \
+    while (stream.get(c)) { \
+      if (c == suffixUpper || c == suffixLower) \
+        break; \
+      if (c >= '0' && c <= '9' || c == '.' || c == '-') \
+        buf.push_back(c); \
+      else \
+        return false; \
+    } \
+    std::stringstream bufStream(buf); \
+    readType value_; \
+    bufStream >> value_; \
+    if (bufStream.get() != EOF) \
+      return false; \
+    if (!stream) \
+      return false; \
+    value = (decltype(value)) value_; \
+    return true; \
+  }
+
+
+rd_snbt_suffix(ByteTag, int, 'B', 'b');
+wr_snbt(ByteTag) { stream << (int) value << 'b'; }  // C++ cannot distinguish byte from char
+
+rd_snbt_suffix(ShortTag, int16_t, 'S', 's');
+wr_snbt(ShortTag) { stream << value << 's'; }
+
+rd_snbt(IntTag) { stream >> value; return !!stream; }  // JavaScript moment
+wr_snbt(IntTag) { stream << value; }
+
+rd_snbt_suffix(LongTag, int64_t, 'L', 'l');
+wr_snbt(LongTag) { stream << value << 'l'; }
+
+rd_snbt_suffix(FloatTag, float, 'F', 'f');
+wr_snbt(FloatTag) { stream.setf(std::ios::fixed, std::ios::floatfield); stream.setf(std::ios::showpoint); stream << std::setprecision(std::numeric_limits<decltype(value)>::max_digits10) << value << 'f'; }
+
+rd_snbt(DoubleTag) { stream >> value; return !!stream; }
+wr_snbt(DoubleTag) { stream.setf(std::ios::floatfield, std::ios::floatfield); stream.setf(std::ios::showpoint); stream << std::setprecision(std::numeric_limits<decltype(value)>::max_digits10) << value; }
```

Comment:
The first call to `stream.setf` should not use `std::ios::floatfield` as first argument. You should be using `std::ios::fixed` here as well, as otherwise doubles are emitted in hexadecimal notation, which Minecraft is unable to parse.